Source/Diagnostics/FullDiagnostics.cpp

```diff
@@ -177,145 +177,138 @@ FullDiagnostics::InitializeFieldFunctorsRZopenPMD (int lev)
             warpx.get_pointer_current_fp(lev, dim)->nComp() == ncomp_multimodefab );
     }
 
-    amrex::Vector< std::string > temp_m_varnames(m_varnames);
-    // Clear any pre-existing vector to release stored data.
-    m_varnames.clear();
-    m_all_field_functors[lev].clear();
-
     // Species index to loop over species that dump rho per species
     int i = 0;
-    bool convert2cartesian;
-    int ncomp;
-    if (m_format == "openpmd") {
-        convert2cartesian = false;
-        ncomp = ncomp_multimodefab;
-    } else {
-        convert2cartesian = true;
-        ncomp = 1;
-    }
-    int n_rz = ncomp * temp_m_varnames.size();
-    m_varnames.reserve(n_rz);
-    m_all_field_functors[lev].resize(temp_m_varnames.size());
+    int ncomp = ncomp_multimodefab;
+    // This function is called multiple times, for different values of `lev`
+    // but the `varnames` need only be updated once.
+    bool update_varnames = (lev==0);
+    if (update_varnames) {
+        m_varnames.clear();
+        int n_rz = ncomp * m_varnames.size();
+        m_varnames.reserve(n_rz);
+    }
+
+    // Reser field functors
+    m_all_field_functors[lev].clear();
+    m_all_field_functors[lev].resize(m_varnames_fields.size());
 
     // Fill vector of functors for all components except individual cylindrical modes.
-    for (int comp=0, n=temp_m_varnames.size(); comp<n; comp++){
-        if        ( temp_m_varnames[comp] == "Er" ){
+    for (int comp=0, n=m_varnames_fields.size(); comp<n; comp++){
```

Comment:
Here we loop over `m_varnames_fields` instead of `m_varnames`, in order to avoid issues when this function is called a second time, **after** `m_varnames` was updated.